File: wini source/evaluador.py
# evaluador.py
"""Evaluación de expresiones en el intérprete wini"""

from exepciones import *
def formatear_texto(valor):
    """Convierte valores a texto para escribir e interpolación."""
    if valor is None:
        return "nulo"
    if isinstance(valor, bool):
        return "verdadero" if valor else "falso"
    if isinstance(valor, str):
        return valor
    if isinstance(valor, list):
        elementos_formateados = [formatear_texto(elem) for elem in valor]
        return "[" + ", ".join(elementos_formateados) + "]"
    if isinstance(valor, dict):
        pares = []
        for k, v in valor.items():
            pares.append(f"{formatear_texto(k)}: {formatear_texto(v)}")
        return "{" + ", ".join(pares) + "}"
    return str(valor)

def es_verdadero(valor):
    """Convierte cualquier valor a booleano según reglas típicas."""
    if isinstance(valor, bool):
        return valor
    if isinstance(valor, (int, float)):
        return valor != 0
    if isinstance(valor, str):
        return len(valor) > 0
    if isinstance(valor, list):
        return len(valor) > 0
    if isinstance(valor, dict):
        return len(valor) > 0
    if valor is None:
        return False
    if hasattr(valor, '__len__'):
        return len(valor) > 0
    return True

def contiene_interpolacion(texto):
    """Verifica si un texto contiene interpolaciones {}"""
    return "{" in texto and "}" in texto

def interpolar_cadena(texto, evaluar_fragmento_func):
    """Interpola variables en una cadena de texto."""
    resultado = []
    indice = 0
    while indice < len(texto):
        caracter = texto[indice]
        if caracter == "{":
            cierre = texto.find("}", indice + 1)
            if cierre == -1:
                from exepciones import RuntimeError
                raise RuntimeError("Cadena interpolada sin cerrar: falta '}'")
            expresion = texto[indice + 1:cierre].strip()
            if not expresion:
                from exepciones import RuntimeError
                raise RuntimeError("Interpolación vacía dentro de '{}'")
            valor = evaluar_fragmento_func(expresion)
            resultado.append(formatear_texto(valor))
            indice = cierre + 1
        else:
            resultado.append(caracter)
            indice += 1
    return "".join(resultado)

class Evaluador:
    """Encapsula la lógica de evaluación de expresiones"""
    
    def __init__(self, interprete):
        self.interprete = interprete
        # Tabla de despacho: tipo de nodo -> método manejador.
        # Se construye una sola vez para no pagar el costo de recorrer una
        # cadena de if/elif (hasta ~17 comparaciones de string) en CADA
        # nodo evaluado. evaluar() es la función más caliente del intérprete
        # (se llama recursivamente por cada expresión, en cada iteración de
        # cada bucle), así que pasar de O(n) a O(1) por nodo aquí es la
        # optimización de mayor impacto posible.
        self._dispatch = {
            "ENTERO": self._ev_entero,
            "DECIMAL": self._ev_decimal,
            "NINGUNO": self._ev_ninguno,
            "CADENA_TEXTO": self._ev_cadena_texto,
            "CADENA_INTERPOLADA": self._ev_cadena_interpolada,
            "BOOLEANO": self._ev_booleano,
            "LISTA": self._ev_lista,
            "DICCIONARIO": self._ev_diccionario,
            "ACCESO_ATRIBUTO": self._ev_acceso_atributo,
            "LLAMADA": self._ev_llamada,
            "INDEXACION": self._ev_indexacion,
            "LLAMADA_METODO": self._ev_llamada_metodo,
            "IDENTIFICADOR": self._ev_identificador,
            "BINARIA": self._evaluar_binaria,
            "UNARIA": self._ev_unaria,
            "LOGICO": self._evaluar_logica,
            "IN": self._ev_in,
        }
    
    def evaluar_fragmento(self, codigo):
        """Evalúa un fragmento de código (usado para interpolación)"""
        from lexer import lexer
        from parser import Parser
        
        tokens_fragmento = lexer(codigo)
        parser_fragmento = Parser(tokens_fragmento)
        ast_fragmento = parser_fragmento.parse()
        if not ast_fragmento.hijos:
            return ""
        return self.evaluar(ast_fragmento.hijos[0])
    
    def evaluar(self, nodo):
        """Evalúa un nodo del AST y retorna su valor.

        Despacha por diccionario (O(1)) en vez de una cadena de if/elif
        (O(n) en el número de tipos de nodo). Como este método es el más
        invocado de todo el intérprete (una vez por cada subexpresión, en
        cada iteración de cada bucle), este cambio es el que más impacta
        en el rendimiento global."""
        manejador = self._dispatch.get(nodo.tipo)
        if manejador is not None:
            return manejador(nodo)
        return None

    # ---------- Manejadores por tipo de nodo (usados por self._dispatch) ----------

    def _ev_entero(self, nodo):
        return int(nodo.valor)

    def _ev_decimal(self, nodo):
        return float(nodo.valor)

    def _ev_ninguno(self, nodo):
        return None

    def _ev_cadena_texto(self, nodo):
        return self._procesar_escapes(str(nodo.valor))

    def _ev_cadena_interpolada(self, nodo):
        texto_procesado = self._procesar_escapes(nodo.valor)
        return interpolar_cadena(texto_procesado, self.evaluar_fragmento)

    def _ev_booleano(self, nodo):
        return nodo.valor

    def _ev_lista(self, nodo):
        return [self.evaluar(elem) for elem in nodo.hijos]

    def _ev_diccionario(self, nodo):
        resultado = {}
        for par in nodo.hijos:
            clave = self.evaluar(par.hijos[0])
            valor = self.evaluar(par.hijos[1])
            resultado[clave] = valor
        return resultado

    def _ev_acceso_atributo(self, nodo):
        atributo_nombre = nodo.valor
        objeto = self.evaluar(nodo.hijos[0])

        # Si es un módulo
        if isinstance(objeto, dict) and objeto.get("_tipo") == "modulo":
            # Buscar en funciones del módulo
            if atributo_nombre in objeto.get("funciones", {}):
                return objeto["funciones"][atributo_nombre]
            # Buscar en variables del módulo
            if atributo_nombre in objeto.get("variables", {}):
                return objeto["variables"][atributo_nombre]
            raise RuntimeError(f"Módulo no tiene '{atributo_nombre}'", linea=nodo.linea)

        raise RuntimeError(f"El valor de tipo '{type(objeto).__name__}' no tiene atributo '{atributo_nombre}'", linea=nodo.linea)

    def _ev_llamada(self, nodo):
        return self.interprete.ejecutor.ejecutar_funcion(nodo)

    def _ev_indexacion(self, nodo):
        objeto = self.evaluar(nodo.hijos[0])
        indice = self.evaluar(nodo.hijos[1])

        if isinstance(objeto, list):
            if not isinstance(indice, int):
                raise ErrorTipo("Índice de lista debe ser entero", linea=nodo.linea)
            if indice < 0 or indice >= len(objeto):
                raise ErrorIndice(f"Índice {indice} fuera de rango para lista de tamaño {len(objeto)}", linea=nodo.linea)
            return objeto[indice]

        elif isinstance(objeto, dict):
            if indice not in objeto:
                raise ErrorIndice(f"Clave '{indice}' no encontrada en el diccionario", linea=nodo.linea)
            return objeto[indice]

        elif isinstance(objeto, str):
            if not isinstance(indice, int):
                raise ErrorTipo("Índice de cadena debe ser entero", linea=nodo.linea)
            if indice < 0 or indice >= len(objeto):
                raise ErrorIndice(f"Índice {indice} fuera de rango para cadena de tamaño {len(objeto)}", linea=nodo.linea)
            return objeto[indice]

        else:
            raise ErrorTipo(f"No se puede indexar sobre tipo {type(objeto).__name__}", linea=nodo.linea)

    def _ev_llamada_metodo(self, nodo):
        return self.interprete.ejecutor.ejecutar_metodo(nodo)

    def _ev_identificador(self, nodo):
        variables = self.interprete.variables
        if nodo.valor in variables:
            return variables[nodo.valor]
        raise RuntimeError(f"Variable '{nodo.valor}' no definida", linea=nodo.linea)

    def _ev_unaria(self, nodo):
        valor = self.evaluar(nodo.hijos[0])
        try:
            return -valor
        except TypeError:
            raise ErrorTipo(
                f"No se puede aplicar '-' unario sobre tipo {type(valor).__name__}",
                linea=nodo.linea
            )

    def _ev_in(self, nodo):
        izquierda = self.evaluar(nodo.hijos[0])
        derecha = self.evaluar(nodo.hijos[1])
        if isinstance(derecha, list):
            return izquierda in derecha
        elif isinstance(derecha, str):
            return str(izquierda) in derecha
        else:
            raise RuntimeError(f"Operador 'in' no soportado entre {type(izquierda)} y {type(derecha)}", linea=nodo.linea)
    
    def _evaluar_binaria(self, nodo):
        """Evalúa operaciones binarias con manejo de errores"""
        try:
            izq = self.evaluar(nodo.hijos[0])
            der = self.evaluar(nodo.hijos[1])
            op = nodo.valor
        except Exception as e:
            raise RuntimeError(f"Error al evaluar operandos: {e}", linea=nodo.linea)
        
        try:
            if op == "+":
                if isinstance(izq, str) or isinstance(der, str):
                    return str(izq) + str(der)
                return izq + der
                
            elif op == "-":
                return izq - der
            elif op == "*":
                if isinstance(izq, str) and isinstance(der, int):
                    return izq * der
                return izq * der
            elif op == "/":
                if der == 0:
                    raise ErrorMatematico("División por cero", linea=nodo.linea)
                return izq / der
            elif op == "%":
                if der == 0:
                    raise ErrorMatematico("Módulo por cero", linea=nodo.linea)
                return izq % der
            elif op == "==":
                return izq == der
            elif op == "!=":
                return izq != der
            elif op == "<>":
                return izq != der
            elif op == "<":
                return izq < der
            elif op == ">":
                return izq > der
            elif op == "<=":
                return izq <= der
            elif op == ">=":
                return izq >= der
            else:
                raise RuntimeError(f"Operador binario desconocido: {op}", linea=nodo.linea)
        
        except ErrorMatematico:
            raise
        except TypeError as e:
            raise ErrorTipo(f"No se puede aplicar operador '{op}' entre {type(izq).__name__} y {type(der).__name__}", linea=nodo.linea)
        except ZeroDivisionError:
            raise ErrorMatematico("División por cero", linea=nodo.linea)
        except Exception as e:
            raise RuntimeError(f"Error en operación {op}: {e}", linea=nodo.linea)
    
    def _evaluar_logica(self, nodo):
        """Evalúa operaciones lógicas"""
        op = nodo.valor
        if op in ("no",):
            valor = self.evaluar(nodo.hijos[0])
            return not es_verdadero(valor)
        elif op in ("y",):
            izquierda = self.evaluar(nodo.hijos[0])
            if not es_verdadero(izquierda):
                return False
            derecha = self.evaluar(nodo.hijos[1])
            return es_verdadero(derecha)
        elif op in ("o",):
            izquierda = self.evaluar(nodo.hijos[0])
            if es_verdadero(izquierda):
                return True
            derecha = self.evaluar(nodo.hijos[1])
            return es_verdadero(derecha)
        else:
            raise RuntimeError(f"Operador lógico desconocido: {op}", linea=nodo.linea)
    

    def _procesar_escapes(self, texto):
        """Procesa caracteres de escape en cadenas de texto."""
        escapes = {
            '\\n': '\n',
            '\\t': '\t',
            '\\r': '\r',
            '\\\\': '\\',
            '\\"': '"',
            "\\'": "'",
        }
        
        resultado = []
        i = 0
        while i < len(texto):
            if texto[i] == '\\' and i + 1 < len(texto):
                escape = texto[i] + texto[i + 1]
                if escape in escapes:
                    resultado.append(escapes[escape])
                    i += 2
                    continue
            resultado.append(texto[i])
            i += 1
        
        return ''.join(resultado)

File: wini/evaluador.go
package wini

import (
	"fmt"
	"strconv"
	"strings"
)

// Evaluador encapsula la lógica de evaluación de expresiones
type Evaluador struct {
	interprete *Interprete // Referencia al intérprete principal
	dispatch   map[string]func(*Nodo) interface{}
}

// NewEvaluador crea un nuevo evaluador
func NewEvaluador(interprete *Interprete) *Evaluador {
	e := &Evaluador{
		interprete: interprete,
		dispatch:   make(map[string]func(*Nodo) interface{}),
	}
	e.initDispatch()
	return e
}

// initDispatch inicializa el mapa de despacho
func (e *Evaluador) initDispatch() {
	e.dispatch["ENTERO"] = e.evEntero
	e.dispatch["DECIMAL"] = e.evDecimal
	e.dispatch["NINGUNO"] = e.evNinguno
	e.dispatch["CADENA_TEXTO"] = e.evCadenaTexto
	e.dispatch["CADENA_INTERPOLADA"] = e.evCadenaInterpolada
	e.dispatch["BOOLEANO"] = e.evBooleano
	e.dispatch["LISTA"] = e.evLista
	e.dispatch["DICCIONARIO"] = e.evDiccionario
	e.dispatch["ACCESO_ATRIBUTO"] = e.evAccesoAtributo
	e.dispatch["LLAMADA"] = e.evLlamada
	e.dispatch["INDEXACION"] = e.evIndexacion
	e.dispatch["LLAMADA_METODO"] = e.evLlamadaMetodo
	e.dispatch["IDENTIFICADOR"] = e.evIdentificador
	e.dispatch["BINARIA"] = e.evaluarBinaria
	e.dispatch["UNARIA"] = e.evUnaria
	e.dispatch["LOGICO"] = e.evaluarLogica
	e.dispatch["IN"] = e.evIn
}

// Evaluar evalúa un nodo del AST y retorna su valor
func (e *Evaluador) Evaluar(nodo *Nodo) interface{} {
	if nodo == nil {
		return nil
	}

	manejador, ok := e.dispatch[nodo.Tipo]
	if !ok {
		return nil
	}
	return manejador(nodo)
}

// FormatearTexto convierte valores a texto para escribir e interpolación
func FormatearTexto(valor interface{}) string {
	if valor == nil {
		return "nulo"
	}

	switch v := valor.(type) {
	case bool:
		if v {
			return "verdadero"
		}
		return "falso"
	case string:
		return v
	case []interface{}:
		elementos := make([]string, len(v))
		for i, elem := range v {
			elementos[i] = FormatearTexto(elem)
		}
		return "[" + strings.Join(elementos, ", ") + "]"
	case map[string]interface{}:
		pares := make([]string, 0, len(v))
		for k, val := range v {
			pares = append(pares, fmt.Sprintf("%s: %s", FormatearTexto(k), FormatearTexto(val)))
		}
		return "{" + strings.Join(pares, ", ") + "}"
	case int, int64, float64:
		return fmt.Sprintf("%v", v)
	default:
		return fmt.Sprintf("%v", v)
	}
}

// EsVerdadero convierte cualquier valor a booleano según reglas típicas
func EsVerdadero(valor interface{}) bool {
	if valor == nil {
		return false
	}

	switch v := valor.(type) {
	case bool:
		return v
	case int:
		return v != 0
	case int64:
		return v != 0
	case float64:
		return v != 0
	case string:
		return len(v) > 0
	case []interface{}:
		return len(v) > 0
	case map[string]interface{}:
		return len(v) > 0
	default:
		return true
	}
}

// ContieneInterpolacion verifica si un texto contiene interpolaciones {}
func ContieneInterpolacion(texto string) bool {
	return strings.Contains(texto, "{") && strings.Contains(texto, "}")
}

// InterpolarCadena interpola variables en una cadena de texto
func (e *Evaluador) InterpolarCadena(texto string, evaluarFragmentoFunc func(string) interface{}) string {
	var resultado strings.Builder
	indice := 0

	for indice < len(texto) {
		caracter := texto[indice]
		if caracter == '{' {
			cierre := strings.Index(texto[indice+1:], "}")
			if cierre == -1 {
				linea := 0
				panic(&RuntimeError{
					ErrorConLinea: ErrorConLinea{
						Mensaje: "Cadena interpolada sin cerrar: falta '}'",
						Linea:   &linea,
					},
				})
			}
			cierre += indice + 1
			expresion := strings.TrimSpace(texto[indice+1 : cierre])
			if expresion == "" {
				linea := 0
				panic(&RuntimeError{
					ErrorConLinea: ErrorConLinea{
						Mensaje: "Interpolación vacía dentro de '{}'",
						Linea:   &linea,
					},
				})
			}
			valor := evaluarFragmentoFunc(expresion)
			resultado.WriteString(FormatearTexto(valor))
			indice = cierre + 1
		} else {
			resultado.WriteByte(caracter)
			indice++
		}
	}
	return resultado.String()
}

// EvaluarFragmento evalúa un fragmento de código (usado para interpolación)
func (e *Evaluador) EvaluarFragmento(codigo string) interface{} {
	lexer := NewLexer()
	tokens := lexer.Tokenize(codigo)
	parser := NewParser(tokens)
	ast := parser.Parse()
	if ast == nil || len(ast.Hijos) == 0 {
		return ""
	}
	return e.Evaluar(ast.Hijos[0])
}

// ========== Manejadores por tipo de nodo ==========

func (e *Evaluador) evEntero(nodo *Nodo) interface{} {
	val, _ := strconv.Atoi(nodo.Valor.(string))
	return val
}

func (e *Evaluador) evDecimal(nodo *Nodo) interface{} {
	val, _ := strconv.ParseFloat(nodo.Valor.(string), 64)
	return val
}

func (e *Evaluador) evNinguno(nodo *Nodo) interface{} {
	return nil
}

func (e *Evaluador) evCadenaTexto(nodo *Nodo) interface{} {
	return e.procesarEscapes(nodo.Valor.(string))
}

func (e *Evaluador) evCadenaInterpolada(nodo *Nodo) interface{} {
	textoProcesado := e.procesarEscapes(nodo.Valor.(string))
	return e.InterpolarCadena(textoProcesado, e.EvaluarFragmento)
}

func (e *Evaluador) evBooleano(nodo *Nodo) interface{} {
	return nodo.Valor
}

func (e *Evaluador) evLista(nodo *Nodo) interface{} {
	resultado := make([]interface{}, len(nodo.Hijos))
	for i, elem := range nodo.Hijos {
		resultado[i] = e.Evaluar(elem)
	}
	return resultado
}

func (e *Evaluador) evDiccionario(nodo *Nodo) interface{} {
	resultado := make(map[string]interface{})
	for _, par := range nodo.Hijos {
		if len(par.Hijos) >= 2 {
			clave := e.Evaluar(par.Hijos[0])
			valor := e.Evaluar(par.Hijos[1])
			// Convertir clave a string para usar como key en el mapa
			resultado[FormatearTexto(clave)] = valor
		}
	}
	return resultado
}

func (e *Evaluador) evAccesoAtributo(nodo *Nodo) interface{} {
	atributoNombre := nodo.Valor.(string)
	objeto := e.Evaluar(nodo.Hijos[0])

	if mapa, ok := objeto.(map[string]interface{}); ok {
		if valor, ok := mapa[atributoNombre]; ok {
			return valor
		}

		if tipo, ok := mapa["_tipo"].(string); ok && tipo == "modulo" {
			if funciones, ok := mapa["funciones"].(map[string]interface{}); ok {
				if fn, ok := funciones[atributoNombre]; ok {
					return fn
				}
			}
			if variables, ok := mapa["variables"].(map[string]interface{}); ok {
				if val, ok := variables[atributoNombre]; ok {
					return val
				}
			}
			linea := 0
			panic(&RuntimeError{
				ErrorConLinea: ErrorConLinea{
					Mensaje: fmt.Sprintf("Módulo no tiene '%s'", atributoNombre),
					Linea:   &linea,
				},
			})
		}
	}

	linea := 0
	panic(&RuntimeError{
		ErrorConLinea: ErrorConLinea{
			Mensaje: fmt.Sprintf("El valor de tipo '%T' no tiene atributo '%s'", objeto, atributoNombre),
			Linea:   &linea,
		},
	})
}

func (e *Evaluador) evLlamada(nodo *Nodo) interface{} {
	return e.interprete.Ejecutor.EjecutarFuncion(nodo)
}

func (e *Evaluador) evIndexacion(nodo *Nodo) interface{} {
	objeto := e.Evaluar(nodo.Hijos[0])
	indice := e.Evaluar(nodo.Hijos[1])
	linea := nodo.Linea

	switch obj := objeto.(type) {
	case []interface{}:
		idx, ok := indice.(int)
		if !ok {
			panic(&ErrorTipo{
				ErrorConLinea: ErrorConLinea{
					Mensaje: "Índice de lista debe ser entero",
					Linea:   &linea,
				},
			})
		}
		if idx < 0 || idx >= len(obj) {
			panic(&ErrorIndice{
				ErrorConLinea: ErrorConLinea{
					Mensaje: fmt.Sprintf("Índice %d fuera de rango para lista de tamaño %d", idx, len(obj)),
					Linea:   &linea,
				},
			})
		}
		return obj[idx]

	case map[string]interface{}:
		clave := FormatearTexto(indice)
		if val, ok := obj[clave]; ok {
			return val
		}
		panic(&ErrorIndice{
			ErrorConLinea: ErrorConLinea{
				Mensaje: fmt.Sprintf("Clave '%s' no encontrada en el diccionario", clave),
				Linea:   &linea,
			},
		})

	case string:
		idx, ok := indice.(int)
		if !ok {
			panic(&ErrorTipo{
				ErrorConLinea: ErrorConLinea{
					Mensaje: "Índice de cadena debe ser entero",
					Linea:   &linea,
				},
			})
		}
		if idx < 0 || idx >= len(obj) {
			panic(&ErrorIndice{
				ErrorConLinea: ErrorConLinea{
					Mensaje: fmt.Sprintf("Índice %d fuera de rango para cadena de tamaño %d", idx, len(obj)),
					Linea:   &linea,
				},
			})
		}
		return string(obj[idx])

	default:
		panic(&ErrorTipo{
			ErrorConLinea: ErrorConLinea{
				Mensaje: fmt.Sprintf("No se puede indexar sobre tipo %T", objeto),
				Linea:   &linea,
			},
		})
	}
}

func (e *Evaluador) evLlamadaMetodo(nodo *Nodo) interface{} {
	return e.interprete.Ejecutor.EjecutarMetodo(nodo)
}

func (e *Evaluador) evIdentificador(nodo *Nodo) interface{} {
	if val, ok := e.interprete.Variables[nodo.Valor.(string)]; ok {
		return val
	}
	linea := nodo.Linea
	panic(&RuntimeError{
		ErrorConLinea: ErrorConLinea{
			Mensaje: fmt.Sprintf("Variable '%s' no definida", nodo.Valor),
			Linea:   &linea,
		},
	})
}

func (e *Evaluador) evUnaria(nodo *Nodo) interface{} {
	valor := e.Evaluar(nodo.Hijos[0])
	linea := nodo.Linea

	switch v := valor.(type) {
	case int:
		return -v
	case int64:
		return -v
	case float64:
		return -v
	default:
		panic(&ErrorTipo{
			ErrorConLinea: ErrorConLinea{
				Mensaje: fmt.Sprintf("No se puede aplicar '-' unario sobre tipo %T", valor),
				Linea:   &linea,
			},
		})
	}
}

func (e *Evaluador) evIn(nodo *Nodo) interface{} {
	izquierda := e.Evaluar(nodo.Hijos[0])
	derecha := e.Evaluar(nodo.Hijos[1])
	linea := nodo.Linea

	switch d := derecha.(type) {
	case []interface{}:
		for _, elem := range d {
			if elem == izquierda {
				return true
			}
		}
		return false
	case string:
		return strings.Contains(d, FormatearTexto(izquierda))
	default:
		panic(&RuntimeError{
			ErrorConLinea: ErrorConLinea{
				Mensaje: fmt.Sprintf("Operador 'in' no soportado entre %T y %T", izquierda, derecha),
				Linea:   &linea,
			},
		})
	}
}

func (e *Evaluador) evaluarBinaria(nodo *Nodo) interface{} {
	linea := nodo.Linea

	var izq, der interface{}
	var err error

	// Evaluar operandos con recuperación de pánico
	func() {
		defer func() {
			if r := recover(); r != nil {
				if e, ok := r.(error); ok {
					err = e
				} else {
					err = fmt.Errorf("%v", r)
				}
			}
		}()
		izq = e.Evaluar(nodo.Hijos[0])
		der = e.Evaluar(nodo.Hijos[1])
	}()

	if err != nil {
		panic(&RuntimeError{
			ErrorConLinea: ErrorConLinea{
				Mensaje: fmt.Sprintf("Error al evaluar operandos: %v", err),
				Linea:   &linea,
			},
		})
	}

	op := nodo.Valor

	// Función auxiliar para manejar pánicos en operaciones
	defer func() {
		if r := recover(); r != nil {
			if _, ok := r.(*ErrorMatematico); ok {
				panic(r)
			}
			if _, ok := r.(*ErrorTipo); ok {
				panic(r)
			}
			panic(&RuntimeError{
				ErrorConLinea: ErrorConLinea{
					Mensaje: fmt.Sprintf("Error en operación %s: %v", op, r),
					Linea:   &linea,
				},
			})
		}
	}()

	// Convertir tipos para operaciones numéricas
	izqFloat, izqIsFloat := toFloat64(izq)
	derFloat, derIsFloat := toFloat64(der)
	izqInt, _ := toInt64(izq)
	derInt, _ := toInt64(der)

	switch op {
	case "+":
		// Concatenación de strings
		if _, ok := izq.(string); ok {
			return fmt.Sprintf("%v%v", izq, der)
		}
		if _, ok := der.(string); ok {
			return fmt.Sprintf("%v%v", izq, der)
		}
		// Suma numérica
		if izqIsFloat || derIsFloat {
			return izqFloat + derFloat
		}
		return int(izqInt + derInt)

	case "-":
		if izqIsFloat || derIsFloat {
			return izqFloat - derFloat
		}
		return int(izqInt - derInt)

	case "*":
		if izqIsFloat || derIsFloat {
			return izqFloat * derFloat
		}
		return int(izqInt * derInt)

	case "/":
		if derIsFloat && derFloat == 0 {
			panic(&ErrorMatematico{
				ErrorConLinea: ErrorConLinea{
					Mensaje: "División por cero",
					Linea:   &linea,
				},
			})
		}
		if !derIsFloat && derInt == 0 {
			panic(&ErrorMatematico{
				ErrorConLinea: ErrorConLinea{
					Mensaje: "División por cero",
					Linea:   &linea,
				},
			})
		}
		if izqIsFloat || derIsFloat {
			return izqFloat / derFloat
		}
		return float64(izqInt) / float64(derInt)

	case "%":
		if derInt == 0 {
			panic(&ErrorMatematico{
				ErrorConLinea: ErrorConLinea{
					Mensaje: "Módulo por cero",
					Linea:   &linea,
				},
			})
		}
		if izqIsFloat || derIsFloat {
			return int(izqFloat) % int(derFloat)
		}
		return int(izqInt % derInt)

	case "==":
		return valoresIguales(izq, der)
	case "!=", "<>":
		return !valoresIguales(izq, der)
	case "<":
		if izqIsFloat || derIsFloat {
			return izqFloat < derFloat
		}
		return izqInt < derInt
	case ">":
		if izqIsFloat || derIsFloat {
			return izqFloat > derFloat
		}
		return izqInt > derInt
	case "<=":
		if izqIsFloat || derIsFloat {
			return izqFloat <= derFloat
		}
		return izqInt <= derInt
	case ">=":
		if izqIsFloat || derIsFloat {
			return izqFloat >= derFloat
		}
		return izqInt >= derInt

	default:
		panic(&RuntimeError{
			ErrorConLinea: ErrorConLinea{
				Mensaje: fmt.Sprintf("Operador binario desconocido: %s", op),
				Linea:   &linea,
			},
		})
	}
}

func (e *Evaluador) evaluarLogica(nodo *Nodo) interface{} {
	op := nodo.Valor
	linea := nodo.Linea

	switch op {
	case "no":
		valor := e.Evaluar(nodo.Hijos[0])
		return !EsVerdadero(valor)

	case "y":
		izquierda := e.Evaluar(nodo.Hijos[0])
		if !EsVerdadero(izquierda) {
			return false
		}
		derecha := e.Evaluar(nodo.Hijos[1])
		return EsVerdadero(derecha)

	case "o":
		izquierda := e.Evaluar(nodo.Hijos[0])
		if EsVerdadero(izquierda) {
			return true
		}
		derecha := e.Evaluar(nodo.Hijos[1])
		return EsVerdadero(derecha)

	default:
		panic(&RuntimeError{
			ErrorConLinea: ErrorConLinea{
				Mensaje: fmt.Sprintf("Operador lógico desconocido: %s", op),
				Linea:   &linea,
			},
		})
	}
}

func (e *Evaluador) procesarEscapes(texto string) string {
	escapes := map[string]string{
		"\\n":  "\n",
		"\\t":  "\t",
		"\\r":  "\r",
		"\\\\": "\\",
		"\\\"": "\"",
		"\\'":  "'",
	}

	var resultado strings.Builder
	i := 0
	for i < len(texto) {
		if texto[i] == '\\' && i+1 < len(texto) {
			escape := texto[i : i+2]
			if val, ok := escapes[escape]; ok {
				resultado.WriteString(val)
				i += 2
				continue
			}
		}
		resultado.WriteByte(texto[i])
		i++
	}
	return resultado.String()
}

// ========== Funciones auxiliares ==========

func toFloat64(val interface{}) (float64, bool) {
	switch v := val.(type) {
	case int:
		return float64(v), true
	case int64:
		return float64(v), true
	case float64:
		return v, true
	default:
		return 0, false
	}
}

func toInt64(val interface{}) (int64, bool) {
	switch v := val.(type) {
	case int:
		return int64(v), true
	case int64:
		return v, true
	case float64:
		return int64(v), true
	default:
		return 0, false
	}
}

// valoresIguales compara dos valores para '==' tratando los números de forma
// coherente sin importar si son int, int64 o float64 internamente
func valoresIguales(a, b interface{}) bool {
	aFloat, aEsNum := toFloat64(a)
	bFloat, bEsNum := toFloat64(b)
	if aEsNum && bEsNum {
		return aFloat == bFloat
	}
	return a == b
}
